src/components/VerticalScroll/VerticalScroll.tsx

```diff
@@ -0,0 +1,224 @@
+import * as React from "react";
+import { withAdaptivity, AdaptivityProps } from "../../hoc/withAdaptivity";
+import { easeInOutSine } from "../../lib/fx";
+import { useEventListener } from "../../hooks/useEventListener";
+import { useExternRef } from "../../hooks/useExternRef";
+import { HasRef } from "../../types";
+import { classNames } from "../../lib/classNames";
+import { ScrollArrow } from "../ScrollArrow/ScrollArrow";
+import { now } from "../../helpers/polymorphNow";
+import { SCROLL_ONE_FRAME_TIME } from "../../helpers/constants";
+import "./VerticalScroll.css";
+
+interface ScrollContext {
+  scrollElement: HTMLElement | null;
+  scrollAnimationDuration: number;
+  animationQueue: VoidFunction[];
+  getScrollPosition: (currentPosition: number) => number;
+  onScrollToBottomBorder: VoidFunction;
+  onScrollEnd: VoidFunction;
+  onScrollStart: VoidFunction;
+  /**
+   * Начальная высота прокрутки.
+   * В некоторых случаях может отличаться от текущей высоты прокрутки из-за transforms: translate
+   */
+  initialScrollHeight: number;
+}
+
+export type ScrollPositionHandler = (currentPosition: number) => number;
+
+export interface VerticalScrollProps
+  extends React.HTMLAttributes<HTMLDivElement>,
+    AdaptivityProps,
+    HasRef<HTMLDivElement> {
+  /**
+   * Функция для расчета величины прокрутки при клике на левую стрелку.
+   */
+  getScrollToTop?: ScrollPositionHandler;
+  /**
+   * Функция для расчета величины прокрутки при клике на правую стрелку.
+   */
+  getScrollToBottom?: ScrollPositionHandler;
+  showArrows?: boolean | "always";
+  scrollAnimationDuration?: number;
+}
+
+const roundUpElementScrollTop = (el: HTMLElement) => Math.ceil(el.scrollTop);
+
+function doScroll({
+  scrollElement,
+  getScrollPosition,
+  animationQueue,
+  onScrollToBottomBorder,
+  onScrollEnd,
+  onScrollStart,
+  initialScrollHeight,
+  scrollAnimationDuration = SCROLL_ONE_FRAME_TIME,
+}: ScrollContext) {
+  if (!scrollElement || !getScrollPosition) {
+    return;
+  }
+
+  /**
+   * максимальное значение сдвига влево
+   */
+  const maxTop = initialScrollHeight - scrollElement.offsetHeight;
+
+  let startTop = roundUpElementScrollTop(scrollElement);
+  let endTop = getScrollPosition(startTop);
+
+  onScrollStart();
+
+  if (endTop >= maxTop) {
+    onScrollToBottomBorder();
+    endTop = maxTop;
+  }
+
+  const startTime = now();
+
+  (function scroll() {
+    if (!scrollElement) {
+      onScrollEnd();
+      return;
+    }
+
+    const time = now();
+    const elapsed = Math.min((time - startTime) / scrollAnimationDuration, 1);
+
+    const value = easeInOutSine(elapsed);
+
+    const currentTop = startTop + (endTop - startTop) * value;
+    scrollElement.scrollTop = Math.ceil(currentTop);
+
+    if (
+      roundUpElementScrollLeft(scrollElement) !== Math.max(0, endTop) &&
```

Comment:
```suggestion
      roundUpElementScrollTop(scrollElement) !== Math.max(0, endTop) &&
```
Чуть выше забыл изменить функцию

Reply:
Применил твои изменения просто)